63.CUDAInterop/main.cpp

```diff
@@ -0,0 +1,409 @@
+			for (auto i = 0; i < familyProps.size(); i++)
+				if (familyProps[i].queueFlags.hasFlags(IQueue::FAMILY_FLAGS::COMPUTE_BIT))
+				{
+					physDev = *physDevIt;
+					params.queueParams[0].familyIndex = i;
+					break;
+				}
+		}
+		if (!physDev)
+			return logFail("Failed to find any Physical Devices with Compute capable Queue Families!");
+
+		{
+			auto& limits = physDev->getLimits();
+			if (!limits.externalMemoryWin32 || !limits.externalFenceWin32 || !limits.externalSemaphoreWin32)
+				return logFail("Physical device does not support the required extensions");
+			
+			cudaHandler = CCUDAHandler::create(system.get(), smart_refctd_ptr<ILogger>(m_logger));
+			assert(cudaHandler);
+			cudaDevice = cudaHandler->createDevice(smart_refctd_ptr_dynamic_cast<CVulkanConnection>(api), physDev);
+		}
+
+		// logical devices need to be created form physical devices which will actually let us create vulkan objects and use the physical device
+		logicalDevice = physDev->createLogicalDevice(std::move(params));
+		if (!logicalDevice)
+			return logFail("Failed to create a Logical Device!");
+		
+		queue = logicalDevice->getQueue(params.queueParams[0].familyIndex, 0);
+		
+		createResources();
+
+		smart_refctd_ptr<ICPUBuffer> ptx;
+		{
+			ISystem::future_t<smart_refctd_ptr<IFile>> fut;
+			m_system->createFile(fut, "../vectorAdd_kernel.cu", IFileBase::ECF_READ);
+			auto [ptx_, res] = cudaHandler->compileDirectlyToPTX(fut.copy().get(), cudaDevice->geDefaultCompileOptions());
+			ASSERT_SUCCESS_NV(res);
+			ptx = std::move(ptx_);
+		}
+		CUmodule   module;
+		CUfunction kernel;
+		CUstream   stream;
+
+		auto& cu = cudaHandler->getCUDAFunctionTable();
+
+		ASSERT_SUCCESS(cu.pcuModuleLoadDataEx(&module, ptx->getPointer(), 0u, nullptr, nullptr));
+		ASSERT_SUCCESS(cu.pcuModuleGetFunction(&kernel, module, "vectorAdd"));
+		ASSERT_SUCCESS(cu.pcuStreamCreate(&stream, CU_STREAM_NON_BLOCKING));
+
+		launchKernel(kernel, stream);
+
+		ASSERT_SUCCESS(cu.pcuStreamSynchronize(stream));
+		ASSERT_SUCCESS(cu.pcuModuleUnload(module));
+		ASSERT_SUCCESS(cu.pcuStreamDestroy_v2(stream));
+
+		return true;
+	}
+
+	void createResources()
+	{
+		auto& cu = cudaHandler->getCUDAFunctionTable();
+
+		for (auto& buf : cpubuffers)
+			buf = make_smart_refctd_ptr<ICPUBuffer>(size);
+
+		for (auto j = 0; j < 2; j++)
+			for (auto i = 0; i < numElements; i++)
+				reinterpret_cast<float*>(cpubuffers[j]->getPointer())[i] = rand() / float(RAND_MAX);
+
+
+
+		ASSERT_SUCCESS(cudaDevice->createSharedMemory(&mem[0], { .size = size, .alignment = sizeof(float), .location = CU_MEM_LOCATION_TYPE_DEVICE }));
+		ASSERT_SUCCESS(cudaDevice->createSharedMemory(&mem[1], { .size = size, .alignment = sizeof(float), .location = CU_MEM_LOCATION_TYPE_DEVICE }));
+		ASSERT_SUCCESS(cudaDevice->createSharedMemory(&mem[2], { .size = size, .alignment = sizeof(float), .location = CU_MEM_LOCATION_TYPE_DEVICE }));
```

Comment:
a simple comment like "create and allocate CUmem with CUDA and slap it inside a simple IReferenceCounted wrapper"